File: python/serialengine/connection.py
from threading import Thread, Lock
from json import dumps as dictToJson
from json import loads as jsonToDict
from json.decoder import JSONDecodeError
import serial

#################
### CONSTANTS ###
#################

from .constants import ACK, NEWLINE, IMG_MSG_S, IMG_MSG_E
from .constants import IMAGE, TYPE, DATA
from .constants import PORT, TIMEOUT, SIZE
from .constants import STATUS, CLOSING, NAME_CONN
from .constants import MAX_RETRIES
from .constants import BAUD_RATE

###############################################################

########################
### CONNECTION CLASS ###
########################


class connection:
    def __init__(self, port, baud=BAUD_RATE, timeout=TIMEOUT, size=SIZE):
        self.canWrite = True
        self.channels = {}
        self.size = size
        self.timeout = timeout
        self.stopped = False
        self.opened = False
        self.port = port
        self.baud = baud
        self.serial = serial.Serial(port, baudrate=baud, timeout=self.timeout)
        self.lock = Lock()

    def start(self):
        Thread(target=self.__run, args=()).start()
        return self

    def __run(self):
        tmp = ""
        while True:
            if self.stopped:
                self.serial.close()
                return

            tmp += self.serial.readline(self.size).decode()

            if tmp == "READY":
                self.opened = True
                tmp = ""

            if tmp != "":
                data = tmp.split("\n")
                for i in range(len(data)):
                    try:
                        msg = jsonToDict(data[i])
                    except JSONDecodeError:
                        continue

                    self.__cascade(msg[TYPE], msg[DATA])
                    self.channels[msg[TYPE]] = msg[DATA]
                    data[i] = ""

                tmp = "".join(data)

    def __cascade(self, mtype, mdata):
        if mtype == ACK:
            self.canWrite = True
        if mtype == STATUS:
            if mdata == CLOSING:
                self.__close()
        if mtype == NAME_CONN:
            self.name = mdata
        if mtype == IMAGE:
            self.write(ACK, ACK)
        return

    def __close(self):
        self.opened = False
        self.stopped = True

    def get(self, channel):
        with self.lock:
            if channel in self.channels.keys():
                return self.channels[channel]
            return None

    def write(self, channel, data):
        if self.opened:
            with self.lock:
                msg = {TYPE: channel.replace("\n", ""), DATA: data.replace("\n", "")}
                self.serial.write(dictToJson(msg).encode() + NEWLINE)

    def close(self):
        self.write(STATUS, CLOSING)
        self.__close()

File: nodejs/connection.js
'use strict'

const EventEmitter = require('events').EventEmitter;
const inherits = require('util').inherits;
const SerialPort = require('serialport');

/////////////////
/// CONSTANTS ///
/////////////////

const { ACK, NEWLINE } = require('./constants');
const { TYPE, DATA } = require('./constants');
const { TIMEOUT, SIZE } = require('./constants');
const { STATUS, CLOSING } = require('./constants');
const { READY } = require('./constants');

///////////////////////////////////////////////////////////////

////////////////////////
/// CONNECTION CLASS ///
////////////////////////

function connection(port, baud, timeout=TIMEOUT, size=SIZE) {
	EventEmitter.call(this);
	this.port = port;
	this.baud = baud;
	this.canWrite = true;
	this.channels = {};
	this.timeout = timeout;
	this.opened = false;
	this.size = size;
	this.msgBuffer = '';
	this.serial = null;
	this.parser = null;

	this.start = function() {
		return new Promise((resolve, reject) => {
			this.serial =  new SerialPort(this.port, { baudRate: this.baud });
			this.parser = this.serial.pipe(new SerialPort.parsers.Ready({ delimiter: READY }));
			this.parser.on('ready', () => {
				this.__run();
				this.opened = true;
				resolve();
			});
		});
	}

	this.__run = function() {
		this.opened = true;
		this.parser.on('data', (bytes) => {
			this.msgBuffer += bytes.toString();
			if(this.msgBuffer != '' && this.msgBuffer != '\n') {
				var data = this.msgBuffer.split('\n');

				for(var i = 0; i < data.length; i++) {
					try {
						if(data[i] == '') {
							continue;
						}
						var msg = JSON.parse(data[i]);
						this.__cascade(msg[TYPE], msg[DATA]);
						this.channels[msg[TYPE]] = msg[DATA];

						this.emit(msg[TYPE], msg[DATA]);
						this.emit('data', msg);
					} catch(err) {console.log(err)};
				}

			}
		});
		this.serial.on('end', () => {
			this.emit('end');
		});

		this.serial.on('error', (err) => {
			this.emit('warning', err);
		});
	}
	

	this.__cascade = function(mtype, mdata) {
		if(mtype == ACK) {
			this.canWrite = true;
		}
		if(mtype == STATUS) {
			if(mdata == CLOSING) {
				this.__close();
			}
		}
	}

	this.__close = function() {
		this.opened = false;
		this.serial.close();
	}

	/////////////////
	/// INTERFACE ///
	/////////////////

	this.get = function(channel) {
		return this.channels[channel];
	}

	this.write = function(dataType, data) {
		if(!this.opened) {
			new Error('Port not opened');
		}
		var msg = {
			'type': dataType.replace('\n', ''),
			'data': data.replace('\n', '')
		};
		this.serial.write(JSON.stringify(msg) + NEWLINE);
	}

	this.close = function() {
		try {
			this.write(STATUS, CLOSING)
		} catch { }
		this.__close()
	}

}

inherits(connection, EventEmitter);

module.exports = exports = connection;
